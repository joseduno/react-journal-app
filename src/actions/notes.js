import Swal from 'sweetalert2'
import {db} from '../firebase/config'
import { fileUpload } from '../helpers/fileUpload'
import { loadNotes } from '../helpers/loadNotes'
import { types } from '../types/types'


export const startNewNote = () => {
    /* getState es una funcion similar al useSelector */ 
    return async(dispatch, getState) => {
        try {
            const {uid} = getState().auth
            const newNote = {
                title: '',
                body: '',
                date: new Date().getTime(),
            }
            const doc = await db.collection(`${uid}/journal/notes`).add(newNote)
            dispatch(activeNote(doc.id, newNote))
            dispatch(addNewNote(doc.id, newNote))
            
        } catch (error) {
            console.log(error)
        }
    }
}


export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
})


export const addNewNote = (id, note) => ({
    type: types.notesAddNew,
    payload: {id, ...note}
})


export const startLoadingNotes = uid => {
    return async(dispatch) => {
        try {
            const notes = await loadNotes(uid)
            dispatch(setNotes(notes))
            
        } catch (error) {
            console.log(error)
        }
    }
}


export const setNotes = notes => ({
    type: types.notesLoad,
    payload: notes
})


export const startSavingNote = note => {
    return async(dispatch, getState) => {
        try {
            const {uid} = getState().auth
            const noteToFirestore = {...note}
            delete noteToFirestore.id
            if (!noteToFirestore.url) {
                delete noteToFirestore.url
            }
            await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore)
            dispatch(refreshNote(note))
            /* Comentado porque en entorno de pruebas me salta un error
            No se hacer el mock del Swal */
            Swal.fire({
                icon: 'success',
                title: 'Saved!',
                text: note.title
            })
            
        } catch (error) {
            console.log(error)
        }
    }
}


export const refreshNote = note => ({
    type: types.notesUpdated,
    payload: note
})


export const startUploadingFile = file => {
    return async(dispatch, getState) => {
        try {
            const {active:activeNote} = getState().notes
            /* Comentado porque en entorno de pruebas me salta un error
            No se hacer el mock del Swal */
            Swal.fire({
                title: 'Uploading...',
                text: 'Please wait...',
                willOpen: () => {
                    Swal.showLoading()
                }
            })
            const fileUrl = await fileUpload(file)
            dispatch(startSavingNote({...activeNote, url: fileUrl}))
            /* Comentado porque en entorno de pruebas me salta un error
            No se hacer el mock del Swal */
            Swal.close()
            
        } catch (error) {
            console.log(error)
        }
    }
}


export const startDeletingNote = id => {
    return async(dispatch, getState) => {
        try {
            const {uid} = getState().auth
            await db.doc(`${uid}/journal/notes/${id}`).delete()
            dispatch(noteDelete(id))
            
        } catch (error) {
            console.log(error)
        }
    }
}


export const noteDelete = id => ({
    type: types.notesDelete,
    payload: id
})


export const logoutCleaning = () => ({
    type: types.notesLogoutCleaning
})