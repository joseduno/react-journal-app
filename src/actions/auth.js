import Swal from 'sweetalert2'
import { firebase, googleAuthProvider } from "../firebase/config"
import { types } from "../types/types"
import { logoutCleaning } from './notes'
import { finishLoading, startLoading } from "./ui"



/* inicio de sesion con firebase */
export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        dispatch(startLoading())
        // No entiendo este return, se usa para que las pruebas funciones, pero no entiendo por que
        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({user}) => {
                dispatch(login(user.uid, user.displayName))
                dispatch(finishLoading())
            })
            .catch(({message}) => {
                Swal.fire('', message, 'error')
                dispatch(finishLoading())
            })
    }
}


/* inicio de sesion con google */
export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({user}) => {
                const {uid, displayName} = user
                dispatch(login(uid, displayName))
            })
            .catch(({message}) => {
                Swal.fire('', message, 'error')
            })
    }
}

/* registro con firebase */
export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {

        // para registrar
        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then(async({user}) => {

                // para actualizar datos
                await user.updateProfile({
                    displayName: name
                })
                dispatch(login(user.uid, user.displayName))
            })
            .catch(({message}) => {
                Swal.fire('', message, 'error')
            })
    }
}

export const startLogout = () => {
    return async(dispatch) => {
        // para cerrar sesion
        try {
            await firebase.auth().signOut()
            dispatch(logout())
            dispatch(logoutCleaning())
        } catch ({message}) {
            Swal.fire('', message, 'error')
        }
    }
}

/* Estas son funciones que retornan acciones!, las anteriores son funciones para 
procesar informacion y posteriormente con un dispatch, ejecutar funciones que retornan 
una accion */
export const login = (uid, displayName) => ({
    type: types.login, 
    payload: {
        uid,
        displayName
    }
})

export const logout = () => ({
    type: types.logout
})