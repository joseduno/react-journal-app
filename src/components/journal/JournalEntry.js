import React from 'react'
import { useDispatch } from 'react-redux'
import { activeNote } from '../../actions/notes'
import moment from 'moment'


export const JournalEntry = ({id, title, body, date, url}) => {

    const noteDate = moment(date)
    const dispatch = useDispatch()

    const handleViewNote = () => {
        dispatch(activeNote(id, {title, body, date, url}))
    }

    return (
        <div 
            className='journal__entry pointer animate__animated animate__slideInLeft animate__faster'
            onClick={handleViewNote}
        >
            {
                url
                &&
                <div
                    className='journal__entry-picture'
                    style={{
                        backgroundSize: 'cover',
                        backgroundImage: `url(${url})`
                    }}
                ></div>
            }
            <div className='journal__entry-body'>
                <p className='journal__entry-title'>{title}</p>
                <p className='journal__entry-content'>{body}</p>
            </div>
            <div className='journal__entry-date-box'>
                <span> {noteDate.format('dddd')} </span>
                <h4>{noteDate.format('Do')}</h4>
            </div>
        </div>
    )
}
