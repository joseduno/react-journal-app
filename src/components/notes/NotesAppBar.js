import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startSavingNote, startUploadingFile } from '../../actions/notes'
import moment from 'moment'


export const NotesAppBar = () => {

    const {active:note} = useSelector(state => state.notes)
    const dispatch = useDispatch()

    const handleSaveClick = () => {
        dispatch(startSavingNote(note))
    }

    const handlePictureClick = () => {
        document.querySelector('#inputPicture').click()
    }

    const handleFileChange = ({target}) => {
        const file = target.files[0]
        if (file) {
            dispatch(startUploadingFile(file))
        }
    }

    return (
        <div className='notes__appbar'>
            <span>{moment().format('LL')}</span>
            <div>
                <input
                    id='inputPicture'
                    type='file'
                    name='file'
                    style={{display: 'none'}}
                    onChange={handleFileChange}
                />
                <button className='btn' onClick={handlePictureClick}>Picture</button>
                <button className='btn' onClick={handleSaveClick}>Save</button>
            </div>
        </div>
    )
}
