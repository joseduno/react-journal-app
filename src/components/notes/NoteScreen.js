import React, { useEffect, useRef } from 'react'
import { useDispatch } from 'react-redux'
import { useSelector } from 'react-redux'
import { activeNote, startDeletingNote } from '../../actions/notes'
import useForm from '../../hooks/useForm'
import { NotesAppBar } from './NotesAppBar'


export const NoteScreen = () => {

    const dispatch = useDispatch()
    const {active:note} = useSelector(state => state.notes)
    const [formValues, handleChange, reset] = useForm(note)
    const {title, body, url} = formValues

    /*Cuando el cambio en una variables (eg note) no ejecuta otras partes del codigo 
    (eg useForm(note)), conviene crear una referencia y efectos secundarios para 
    poder hacer notar los cambios donde se necesiten, como se muestra a continuacion*/
    const activeId = useRef(note.id)
    useEffect(() => {
        if (note.id !== activeId.current) {
            reset(note)
            activeId.current = note.id
        }
    }, [note, reset])

    useEffect(() => {
        dispatch(activeNote(formValues.id, formValues))
    }, [formValues, dispatch])

    const handleDelete = () => {
        dispatch(startDeletingNote(note.id))
    }
    
    return (
        <div className='notes__main-content animate__animated animate__fadeIn animate__faster'>
            <NotesAppBar />

            <div className='notes__content'>
                <input
                    type='text'
                    placeholder='some awesome title'
                    className='notes__title-input'
                    autoComplete='off'
                    name='title'
                    value={title}
                    onChange={handleChange}
                />

                <textarea
                    placeholder='what happened today?'
                    className='notes__textarea'
                    name='body'
                    value={body}
                    onChange={handleChange}
                >
                </textarea>

                {
                    url
                    &&
                    <div className='notes__image'>
                        <img
                            src={`${url}`}
                            alt='imagen'
                            name='url'
                        />
                    </div>
                }
            </div>
            <button className='btn btn-danger' onClick={handleDelete}>
                Delete
            </button>
        </div>
    )
}
