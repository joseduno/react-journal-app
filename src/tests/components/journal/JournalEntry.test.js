import { mount } from "enzyme"
import { Provider } from "react-redux"
import thunk from "redux-thunk"
import { JournalEntry } from "../../../components/journal/JournalEntry"
import configureStore from 'redux-mock-store'
import { activeNote } from "../../../actions/notes"


jest.mock("../../../actions/notes", () => ({
    activeNote: jest.fn()
}))


const middlewares = [thunk]
const mockStore = configureStore(middlewares)
const initialState = {}
let store = mockStore(initialState)
store.dispatch = jest.fn()


describe('Pruebas en el componente JournalEntry.js', () => {

    const entry = {
        id: 'testId',
        title: 'testTitle',
        body: 'testBody',
        date: 12341234,
        url: 'testUrl'
    } 

    const wrapper = mount(
        <Provider store={store}>
            <JournalEntry {...entry} />
        </Provider>
    )

    beforeEach(() => {
        store = mockStore(initialState)
        jest.clearAllMocks()
    })

    test('Debe mostrarse correctametne', () => {
        expect(wrapper).toMatchSnapshot()
    })

    test('Debe hacer dispatch de activeNote', () => {
        wrapper.find('.journal__entry').simulate('click')
        expect(activeNote).toHaveBeenCalledWith(entry.id, {
            title: entry.title,
            body: entry.body,
            date: entry.date,
            url: entry.url,
        })
    })
})
