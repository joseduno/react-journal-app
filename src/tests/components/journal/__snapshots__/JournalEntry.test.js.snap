// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente JournalEntry.js Debe mostrarse correctametne 1`] = `
<div
  className="journal__entry pointer animate__animated animate__slideInLeft animate__faster"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(testUrl)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      testTitle
    </p>
    <p
      className="journal__entry-content"
    >
      testBody
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
       
      Thursday
       
    </span>
    <h4>
      1st
    </h4>
  </div>
</div>
`;
